### PR TITLE
add AI couplet sample code and README file

### PR DESCRIPTION
**Pull request by TobeyQin**

### CODE REVIEW

**Comment by mslichao** [comment]
You can add a link in the root README.md file.

File: B-教学案例与实践/B13-AI对联生成案例/README.md [hunk 1]
```diff
@@ -0,0 +1,492 @@
...
+
+### 操作系统
+
+本案例运行在Azure虚拟机上，虚拟机的系统为Ubuntu 16.04
+
+### 编程语言
+
+本案例的语言使用 Python3.x ，并需要安装一些 python packages。
+
+1. 训练所需 python packages 在文件 [train_requirements.txt](./src/training/pip_requirerments.txt) 中。
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/README.md]: >pip_requirerments.txt [](start = 68, length = 21)

train_requirerments.txt

File: B-教学案例与实践/B13-AI对联生成案例/README.md [hunk 2]
```diff
@@ -0,0 +1,492 @@
...
+## 模型训练
+
+### 模型选择
+
+想要完成一个自动生成对联的小程序，想法十分美好，但想要达到这个目标，光拍拍脑袋想想是不够的，需要训练出一个能完成对联生成的自然语言理解模型。于是乎，就有两个选择：
+
+1. 自己写一套完成对联生成工作的深度学习模型。这个工作量相当之大，可能需要一个NLP专业团队来进行开发，调优。
+2. 应用已有的深度学习模型，直接应用。这个选择比较符合客观需要。我们找到了两个工具包：
+
+    + Tensor2Tensor 工具包：Tensor2Tensor（以下简称T2T）是由 Google Brain 团队使用和维护的开源深度学习模型库，支持多种数据集和模型。T2T 在 github 上有完整的介绍和用法，可以访问这里了解详细信息。
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/README.md]: >这里 [](start = 119, length = 2)

link?

File: B-教学案例与实践/B13-AI对联生成案例/README.md [hunk 3]
```diff
@@ -0,0 +1,492 @@
...
+
+    + Fairseq 工具包：Fairseq 是 Facebook 推出的一个序列建模工具包，这个工具包允许研究和开发人员自定义训练翻译、摘要、语言模型等文本生成任务。这里是它的 PyTorch 实现。
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/README.md]: >这里 [](start = 87, length = 2)

link?

File: B-教学案例与实践/B13-AI对联生成案例/README.md [hunk 4]
```diff
@@ -0,0 +1,492 @@
...
+    with open(filename, 'r', encoding='utf-8') as infile:
+    with open(filename + '.clean', 'w', encoding='utf-8') as outfile:
+        lines = infile.readlines()
+        for line in lines:
+        out = ""
+        for i in line.strip():
+            out += i + (' ')
+        out = out[:-1]
+        out += '\n'
+        outfile.write(out)
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/README.md]: Seems the indentation not right after paste here.

File: B-教学案例与实践/B13-AI对联生成案例/README.md [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+    b. 执行如下命令完成文件分隔
+    ```
+    python split_data.py train.txt.up
+    python split_data.py train.txt.down
+    ```
+    分隔后生成两个文件：train.txt.up.clean 和 train.txt.down.clean
+
+    c. 合并文件为 merge.txt
+    ```
+    cat train.txt.up train.txt.down > merge.txt
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/README.md]: >train.txt.up train.txt.down [](start = 8, length = 27)

Should we use the .clean files?

File: B-教学案例与实践/B13-AI对联生成案例/src/pai_config/hdfshelper.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import os
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/src/pai_config/hdfshelper.py]: Add license for every code file.

# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.

File: B-教学案例与实践/B13-AI对联生成案例/src/service/synthesis.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    bg_img_2 =  Image.open(background_image_path) # background image
+    bg_img_2_resized = bg_img_2.resize((800,600), Image.ANTIALIAS)
+    bg_crop= np.asarray(np.asarray(bg_img_2_resized))[0:h, 0:w,:]
+
+    mask,mask_2 = gen_cir_mask(w, h)
+    mask =  np.transpose(np.repeat(np.expand_dims(mask,0), 3, axis=0), (1,2,0))
+    mask_2 =  np.transpose(np.repeat(np.expand_dims(mask_2,0), 3, axis=0), (1,2,0))
+
+    #word_font = ImageFont.truetype("arial.ttf", 40)
+    word_font = ImageFont.truetype(u"/home/administor/code/Poetic-Image/imgs/handwriting.ttf", size=35)
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/src/service/synthesis.py]: >/home/administor/code/Poetic-Image/imgs/handwriting.ttf [](start = 37, length = 55)

Seems this file not in the PR

File: B-教学案例与实践/B13-AI对联生成案例/src/service/up2down_class.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                    "__init__.py file should include the necessary imports. "
+                    "The imported files should contain registrations, "
+                    "e.g. @registry.register_model calls, that will then be "
+                    "available to the t2t-decoder.")
+flags.DEFINE_string("master", "", "Address of TensorFlow master.")
+flags.DEFINE_string("schedule", "train_and_evaluate",
+                    "Must be train_and_evaluate for decoding.")
+
+
+home = '/home/administor/code/test_model'
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/src/service/up2down_class.py]: >/home/administor/code/test_model [](start = 8, length = 32)

Can we use a relative path

File: B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    train.in.txt，train.out.txt，test.in.txt，test.out.txt。
+    ```
+3. 在 data 目录中，合并上下联文件并保存结果：
+    ```
+    cat train.in.txt train.out.txt > train.merge.txt
+    ```
+4. 在data 目录中，统计字表数目并保存字表：
+
+    需要安装 subword-nmt 包：
+    ` ``
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md]: >` `` [](start = 4, length = 4)

Remove the space

File: B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    python -m pip install subword-nmt
```
Comment (mslichao) [inline B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md]: >subword-nmt [](start = 26, length = 11)

How about add this package to `train_requirements.txt`

File: B-教学案例与实践/B13-AI对联生成案例/README.md (hunk 1 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: Thanks! Fixed!

File: B-教学案例与实践/B13-AI对联生成案例/README.md (hunk 2 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: added!

File: B-教学案例与实践/B13-AI对联生成案例/README.md (hunk 3 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: added!

File: B-教学案例与实践/B13-AI对联生成案例/README.md (hunk 5 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: both are fine for this operation.

Comment (mslichao) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: If we don't use the .clean file, why we need to run step b?

---
In reply to: [279605297](https://github.com/Microsoft/ai-edu/pull/241#discussion_r279605297) [](ancestors = 279605297)

File: B-教学案例与实践/B13-AI对联生成案例/src/service/synthesis.py (hunk 7 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/src/service/synthesis.py]: added

File: B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md (hunk 9 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md]: fixed

File: B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md (hunk 10 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/src/training/how_to_train.md]: Sure. Added to the requirements file

File: B-教学案例与实践/B13-AI对联生成案例/src/service/up2down_class.py (hunk 8 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/src/service/up2down_class.py]: change to relative path

File: B-教学案例与实践/B13-AI对联生成案例/src/pai_config/hdfshelper.py (hunk 6 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/src/pai_config/hdfshelper.py]: added

File: B-教学案例与实践/B13-AI对联生成案例/README.md (hunk 4 above)
Comment (TobeyQin) [reply B-教学案例与实践/B13-AI对联生成案例/README.md]: fixed

**Pull request merged**